warcprox/mitmproxy.py

```diff
@@ -380,14 +384,31 @@ def do_COMMAND(self):
             else:
                 self._determine_host_port()
                 assert self.url
-
+            # Check if target hostname:port is in `bad_hostnames_ports` cache
+            # to avoid retrying to connect.
+            with self.server.bad_hostnames_ports_lock:
+                hostname_port = self._hostname_port_cache_key()
+                if hostname_port in self.server.bad_hostnames_ports:
+                    self.logger.info('Cannot connect to %s (cache)',
+                                     hostname_port)
+                    self.send_error(502, 'message timed out')
+                    return
```

Comment:
I hope it's not necessary to lock for a read-only operation?

Reply:
@nlevitt read and write operations are NOT thread-safe for `TTLCache`. I double checked that.

Ref: https://cachetools.readthedocs.io/en/stable/
_Please be aware that all these classes are not thread-safe. **Access to a shared cache** from multiple threads must be properly synchronized, e.g. by using one of the memoizing decorators with a suitable lock object._

What is preventing the cache from being thread-safe for read operations is that it calls an `expire()` method to delete old entries and this could raise a `KeyError` when we try to delete the same entry concurrently from multiple threads.
Actually there was an issue about that: https://github.com/tkem/cachetools/issues/115

Reply:
Ah, ok!